server/controllers/images.ts

```diff
@@ -0,0 +1,141 @@
+    return { width, height, blurHash };
+  } else {
+    return null;
+  }
+};
+
+const recordUploadedFile = async (
+  req,
+  kind,
+  file,
+  fileName,
+  s3Data: ManagedUpload.SendData | { Location: string; Bucket: string; Key: string },
+): Promise<void> => {
+  try {
+    await models.UploadedFile.create({
+      kind: kind,
+      fileName,
+      fileSize: file.size,
+      fileType: file.mimetype,
+      url: s3Data.Location,
+      data: await getUploadedFileData(file),
+      CreatedByUserId: req.remoteUser.id,
+    });
+  } catch (e) {
+    // Since this runs async, we need to report the error to Sentry in case it fails
+    console.error(e);
+    reportErrorToSentry(e, {
+      severity: 'error',
+      user: req.remoteUser,
+      transactionName: 'recordUploadedFile',
+      extra: { kind, fileName, fileType: file.mimetype },
+    });
+  }
+};
+
+const getFilename = (file, fileNameFromArgs) => {
+  const expectedExtension = SUPPORTED_FILE_EXTENSIONS[file.mimetype];
+  const rawFileName = fileNameFromArgs || file.originalname || uuid();
+  const parsedFileName = path.parse(rawFileName);
+  return `${parsedFileName.name}${expectedExtension}`;
+};
+
+export default async function uploadImage(req, res, next) {
+  const { kind } = req.body;
+
+  if (!req.remoteUser) {
+    return next(new errors.Unauthorized('You need to be logged in to upload a file'));
+  }
+
+  if (!kind) {
+    return next(
+      new errors.ValidationFailed('missing_required', {
+        kind: 'Kind field is required and missing',
+      }),
+    );
+  }
+
+  if (!SUPPORTED_FILE_KINDS.includes(kind)) {
+    const message = `Kind should be one of: ${SUPPORTED_FILE_KINDS.join(', ')}`;
+    return next(new errors.ValidationFailed('INVALID_FILE_KIND', { kind: message }, message));
+  }
+
+  // Required fields
+  const file = req.file;
+  if (!file) {
+    return next(
+      new errors.ValidationFailed('missing_required', {
+        file: 'File field is required and missing',
+      }),
+    );
+  }
+
+  // Validate file
+  if (!SUPPORTED_FILE_TYPES.includes(file.mimetype)) {
+    const message = `Mimetype of the file should be one of: ${SUPPORTED_FILE_TYPES.join(', ')}`;
+    return next(new errors.ValidationFailed('INVALID_FILE_MIME_TYPE', { file: message }, message));
+  }
+
+  if (file.size > 1024 * 1024 * 10) {
+    const message = 'Filesize cannot exceed 10MB';
+    return next(
+      new errors.ValidationFailed('INVALID_FILE_SIZE_TOO_BIG', { file: message }, message, {
+        fileSize: file.size,
+        max: '10MB',
+      }),
+    );
+  }
+
+  if (!s3) {
+    return next(new errors.ServerError('S3 service object not initialized'));
+  }
+
+  /**
+   * We will replace the name to avoid collisions
+   */
+  const fileName = getFilename(file, req.body.fileName);
+  const uploadParams = {
+    Bucket: config.aws.s3.bucket,
+    Key: `${kebabCase(kind)}/${uuid()}/${fileName || uuid()}`,
```

Comment:
My suggestion is to add the kind to metadata and file name to the content disposition. And to have a single uuid to refer to this file in storage. 

```suggestion
    Key: `uploaded-files/${uuid()}`,
```

Reply:
Storing under the right kind is meant to better organize files when browsing from S3's UIs. What are the concerns with the kind and filename being visible in the path?

Reply:
My reasoning was that the file name could contain PII or inapropriate user generated content by itself.
But having the url means having access to the file, so I dont know if thats would be an issue.